src/charm.py

```diff
@@ -31,12 +31,53 @@ def __init__(self, *args):
 
         self.framework.observe(self.on.install, self._on_install)
         self.framework.observe(self.on.config_changed, self._on_config_changed)
+        self.framework.observe(self.on.mongodb_relation_joined, self._mongodb_relation_joined)
         self.framework.observe(self.on.start, self._on_start)
 
+    def _mongodb_relation_joined(self, event) -> None:
+        if not self.unit.is_leader():
+            return
+
+        # do not initialise replica set until all peers have joined
+        if not self.app.planned_units() == len(self._peers.units) + 1:
+            logger.debug(
+                "planned units does not match joined units: %s != %s, will "
+                + "wait to initialise replica set until all planned units "
+                + "have joined",
+                str(self.app.planned_units()),
+                len(self.model.get_relation(PEER).units) + 1,
+            )
+            return
+
+        # do not initilise replica set until all peers have started mongod
+        for peer in self._peers.units:
+            peer_ip = str(self._peers.data[peer].get("private-address"))
+            peer_config = self._config
+            peer_config["calling_unit_ip"] = peer_ip
+            mongo_peer = MongoDB(peer_config)
+            if not mongo_peer.is_ready(all_replicas=False):
+                logger.debug(
+                    "unit: %s is not ready, cannot initilise replica set "
+                    + "until all units are ready, defering on relation-joined",
+                    peer,
+                )
+                event.defer()
+                return
+
+        # initialise replica set if not already done
+        if not self._mongo.is_replica_set():
```

Comment:
This is also confusing I apologise. The reason we don't block here is because it is possible that the application `MongoDB` could have been deployed with _one unit_ . If it is deployed as _one unit_ then it will initialise the replicaset in [_on_start](https://github.com/canonical/mongodb-operator/blob/41d005df7800f8e6a685e13a771f7b50371e4f7a/src/charm.py#L150-L151)

Reply:
Hi deleted my first response because I misunderstood what you were saying. My apologies!

Reply:
`_initialise_replica_set` will put it into a `BlockedStatus` if it isn't ready but I guess there is a need to handle an exception here so that way it doesn't execute the rest of the code after an initialisation failure. Thanks for catching this @DnPlas